### PR TITLE
feat(*): use enums for consistency and speed

### PR DESCRIPTION
**Pull request by iShibi**
**Please describe the changes this PR makes and why it should be merged:**

Okay, there are a lot of related changes here:

1) Use enums instead of arrays. This brings consistency and speed (`indexOf` is slow).
2) Some enums (that were arrays earlier) had same name for the enums themselves and the type of elements they hold.
3) `WebhookTypes` enum (ex-array) was missing from typings for `Constants`
4) Some methods that used arrays as enums earlier wouldn't let user to pass `number` type.
5) `DefaultMessageNotifcations` is now `DefaultMessageNotificationLevels` to keep the naming consistent with other enums in djs (discord uses 'level' in the name for enums in documentaion).
6) `DefaultMessageNotifcations` types were not same as what the api has documented ( eg. `ALL` instead of `ALL_MESSAGES`)
6) Although, the raw activity data's `name` property's value is `Custom_Status`, the `type` for it is documented as `Custom`. So, the activity type `CUSTOM_STATUS` is now `CUSTOM`.  

**Status and versioning classification:**
- Code changes have been tested against the Discord API
- I know how to update typings and have done so
- This PR includes breaking changes

### CODE REVIEW

File: src/managers/GuildManager.js [hunk 1]
```diff
@@ -163,14 +164,21 @@ class GuildManager extends BaseManager {
     } = {},
   ) {
     icon = await DataResolver.resolveImage(icon);
-    if (typeof verificationLevel !== 'undefined' && typeof verificationLevel !== 'number') {
-      verificationLevel = VerificationLevels.indexOf(verificationLevel);
+    if (verificationLevel) {
+      verificationLevel =
+        typeof verificationLevel === 'number' ? verificationLevel : VerificationLevels[verificationLevel];
```
Comment (kyranet) [inline src/managers/GuildManager.js:169]: You could probably change this to:
```suggestion
    if (typeof verificationLevel === 'string') {
      verificationLevel = VerificationLevels[verificationLevel];
```

And same for the others, those values take either a string or a number, one is untouched, and the other is transformed.

Comment (kyranet) [reply src/managers/GuildManager.js:169]: Same goes fo the other checks.

**Review by kyranet — APPROVED** [review]

**Review by kyranet — CHANGES_REQUESTED** [review]
Sorry, wrong button... ^

**Review by kyranet — APPROVED** [review]

File: src/util/Constants.js [hunk 2]
```diff
@@ -708,38 +708,28 @@ exports.APIErrors = {
...
 
 /**
- * The value set for a guild's default message notifications, e.g. `ALL`. Here are the available types:
- * * ALL
- * * MENTIONS
- * @typedef {string} DefaultMessageNotifications
+ * The value set for a guild's default message notifications, e.g. `ALL_MESSAGES`. Here are the available types:
+ * * ALL_MESSAGES
+ * * ONLY_MENTIONS
+ * @typedef {string} DefaultMessageNotificationLevel
```
Comment (vladfrangu) [inline src/util/Constants.js:714]: Why is this renamed, when the property has the same name

File: src/util/Constants.js [hunk 3]
```diff
@@ -708,38 +708,28 @@ exports.APIErrors = {
...
  */
-exports.DefaultMessageNotifications = ['ALL', 'MENTIONS'];
+exports.DefaultMessageNotificationLevels = createEnum(['ALL_MESSAGES', 'ONLY_MENTIONS']);
 
 /**
  * The value set for a team members's membership state:
  * * INVITED
  * * ACCEPTED
- * @typedef {string} MembershipStates
+ * @typedef {string} MembershipState
```
Comment (vladfrangu) [inline src/util/Constants.js:722]: Why is this renamed, when the property has the same name

File: src/util/Constants.js [hunk 4]
```diff
@@ -708,38 +708,28 @@ exports.APIErrors = {
...
-  'ACCEPTED',
-];
+exports.MembershipStates = createEnum([null, 'INVITED', 'ACCEPTED']);
 
 /**
  * The value set for a webhook's type:
  * * Incoming
  * * Channel Follower
- * @typedef {string} WebhookTypes
+ * @typedef {string} WebhookType
```
Comment (vladfrangu) [inline src/util/Constants.js:730]: Why is this renamed, when the property has the same name

File: src/structures/Guild.js [hunk 5]
```diff
@@ -850,17 +850,17 @@ class Guild extends BaseGuild {
     if (data.splash) _data.splash = data.splash;
     if (data.discoverySplash) _data.discovery_splash = data.discoverySplash;
     if (data.banner) _data.banner = data.banner;
-    if (typeof data.explicitContentFilter !== 'undefined') {
+    if (data.explicitContentFilter) {
```
Comment (vladfrangu) [inline src/structures/Guild.js:853]: This won't match if the filter is 0

File: src/structures/Guild.js [hunk 6]
```diff
@@ -832,11 +832,11 @@ class Guild extends BaseGuild {
   edit(data, reason) {
     const _data = {};
     if (data.name) _data.name = data.name;
-    if (typeof data.verificationLevel !== 'undefined') {
+    if (data.verificationLevel) {
```
Comment (vladfrangu) [inline src/structures/Guild.js:835]: This won't match if the level is 0

File: src/structures/Guild.js [hunk 7]
```diff
@@ -850,17 +850,17 @@ class Guild extends BaseGuild {
...
-    if (typeof data.explicitContentFilter !== 'undefined') {
+    if (data.explicitContentFilter) {
       _data.explicit_content_filter =
         typeof data.explicitContentFilter === 'number'
           ? data.explicitContentFilter
-          : ExplicitContentFilterLevels.indexOf(data.explicitContentFilter);
+          : ExplicitContentFilterLevels[data.explicitContentFilter];
     }
-    if (typeof data.defaultMessageNotifications !== 'undefined') {
+    if (data.defaultMessageNotifications) {
```
Comment (vladfrangu) [inline src/structures/Guild.js:859]: This won't match if the level is 0

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/util/Constants.js (hunk 2 above)
Comment (iShibi) [reply src/util/Constants.js:714]: That is how all the enums are documented. The enum is named plural and the type of elements it stores have the singular name. Eg. enum: `NSFWLevels`, the type it stores: `NSFWLevel`

Comment (vladfrangu) [reply src/util/Constants.js:714]: 👍

File: src/structures/Guild.js (hunk 6 above)
Comment (iShibi) [reply src/structures/Guild.js:835]: yeah, right. I'll fix it 😸

**Review by SpaceEEC — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Comment by iCrawl** [comment]
This needs a rebase.

**Comment by iShibi** [comment]
Rebased
